### PR TITLE
Grid system

### PR DESCRIPTION
**Pull request by eldiegod**
Fixes: #1042

### CODE REVIEW

File: src/shared/components/Grid2/Grid2.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+type ReponsivenessArray = Array<number | null>
+
+type GridProps = {
+  gap?: number | ReponsivenessArray
+  cols?: number | ReponsivenessArray
+}
+
+// Applies styles to the corresponding media queries given by the values passed to the ReponsivenessArray
+const applyStyles = <T extends Record<string, unknown & { toString: () => string }>>(
```
Comment (kdembler) [inline src/shared/components/Grid2/Grid2.tsx]: Can you explain what's going on here? What's the high level explanation of how this component works?

**Comment by eldiegod** [comment]
I.e for the props, values or array of values can be used, each value in the array will be applied to a media query starting from the smaller one so gap={[8,24]} will apply grid-gap: 8px; to media query base and grid-gap: 24px; to media query compact
@kdembler

**Comment by eldiegod** [comment]
the component needs a bit more testing but wanted to have an opinion on the API and about what should we name it since we have a Grid component already (not sure if replaceable)

File: src/shared/components/GridLayout/GridLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import styled from '@emotion/styled'
+
+import { media } from '@/shared/theme'
+
+type ReponsivenessObject = Partial<Record<keyof typeof media, number>>
+
+export const GridLayout = styled.div`
```
Comment (kdembler) [inline src/shared/components/GridLayout/GridLayout.tsx]: Let's call it `LayoutGrid` 🙏

File: src/shared/components/GridLayout/GridLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+`
+
+type GridItemProps = {
+  colStart?: number | 'initial' | ReponsivenessObject
+  colSpan?: number | ReponsivenessObject
+  rowStart?: number | 'initial' | ReponsivenessObject
+  rowSpan?: number | ReponsivenessObject
+}
+
+function isResponsiveObject(prop?: number | 'initial' | ReponsivenessObject): prop is ReponsivenessObject {
```
Comment (kdembler) [inline src/shared/components/GridLayout/GridLayout.tsx]: I guess this should be `isResponsivenessObject`

File: src/shared/components/GridLayout/GridLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  return !!prop && typeof prop !== 'number' && prop !== 'initial'
+}
+
+export const GridItem = styled.div<GridItemProps>`
+  ${media.base} {
+    ${({ colStart }) => isResponsiveObject(colStart) && `grid-column-start: ${colStart.base};`}
+    ${({ colSpan }) => isResponsiveObject(colSpan) && `grid-column-end: span ${colSpan.base};`}
+    ${({ rowStart }) => isResponsiveObject(rowStart) && `grid-row-start: ${rowStart.base};`}
+    ${({ rowSpan }) => isResponsiveObject(rowSpan) && `grid-row-end: span ${rowSpan.base};`}
+  }
```
Comment (kdembler) [inline src/shared/components/GridLayout/GridLayout.tsx]: We can still use something to help create those. How about:

```suggestion

const createBreakpointGridRules = (breakpointKey: keyof ReponsivenessObject) => ({ colStart, colSpan, rowStart, rowSpan }: GridItemProps) => css`
  ${isResponsiveObject(colStart) && `grid-column-start: ${colStart[breakpointKey]};`}
  // ...
`

export const GridItem = styled.div<GridItemProps>`
  ${media.base} {
    ${createBreakpointGridRules('base')}
  }
```

**Review by kdembler — APPROVED** [review]

**Pull request merged**